src/openapi/streaming/streaming.js

```diff
@@ -523,6 +525,103 @@ Streaming.prototype.disposeSubscription = function(subscription) {
     }
 };
 
+function handleSubscriptionReadyForUnsubscribe(subscriptions, resolve) {
+    let allSubscriptionsReady = true;
+    for (let i = 0; i < subscriptions.length && allSubscriptionsReady; i++) {
+        if (!subscriptions[i].isReadyForUnsubscribeByTag()) {
+            allSubscriptionsReady = false;
+        }
+    }
+
+    if (allSubscriptionsReady) {
+        resolve();
+    }
+}
+
+function getSubscriptionsToRemove(serviceGroup, url, tag) {
+    const subscriptionsToRemove = [];
+
+    for (let i = 0; i < this.subscriptions.length; i++) {
+        const subscription = this.subscriptions[i];
+
+        if (subscription.serviceGroup === serviceGroup &&
+                subscription.url === url &&
+                subscription.subscriptionData.Tag === tag) {
+
+            subscriptionsToRemove.push(subscription);
+        }
+    }
+
+    return subscriptionsToRemove;
+}
+
+function unsubscribeSubscriptionByTag(serviceGroup, url, tag, shouldDisposeSubscription) {
+    const subscriptionsToRemove = getSubscriptionsToRemove.call(this, serviceGroup, url, tag);
+
+    let allSubscriptionsReadyResolve;
+    const allSubscriptionsReady = new Promise((resolve) => {

```

Comment:
What do you think about extracting Promise to separate function so We could end up with **unsubscribeSubscriptionByTag** looking like below. I think it would be easier to reason about.
```
function unsubscribeSubscriptionByTag(serviceGroup, url, tag, shouldDisposeSubscription) {
    const subscriptionsToRemove = getSubscriptionsToRemove.call(this, serviceGroup, url, tag);
    const allSubscriptionsReady = getSubscriptionsReadyPromise.call(this, subscriptionsToRemove, onStateChanged, shouldDisposeSubscription);

    allSubscriptionsReady.then(() => {
        ...
    });
}
```